### ISSUE TITLE
PhoBERT with underthesea

### ISSUE THREAD

**Issue by DoDucHuy21020124** [undertheseanlp/underthesea#760]
Can underthesea package be used for the PhoBERT encoder without harming the performance of the encoder?

**Comment by rain1024** [comment]
Short answer: **yes.** PhoBERT expects word-segmented input (syllables of a multi-syllable word joined by `_`), and `underthesea.word_tokenize(text, format="text")` produces exactly that format, so you can feed its output straight into the PhoBERT tokenizer.

PhoBERT was pre-trained on text segmented by VnCoreNLP's RDRSegmenter, so we measured how much switching the segmenter to underthesea actually matters, on one sentence-level task and one sequence-labeling task (`vinai/phobert-base-v2`, underthesea 9.5.0, VnCoreNLP 1.2).

### Sentiment — UIT-VSFC (11,426 train / 3,166 test)

| Segmenter | Frozen encoder + LR, acc | Fine-tuned 3 ep, acc | Fine-tuned, weighted-F1 |
|---|---|---|---|
| VnCoreNLP (RDRSegmenter) | 92.26 | 94.00 | 93.58 |
| **underthesea `word_tokenize`** | 92.17 | 93.97 | 93.59 |
| none (raw syllables) | 90.56 | 93.59 | 93.27 |

### NER — PhoNER_COVID19 (5,027 train / 3,000 test), frozen encoder + linear token classifier, entity-level F1 scored at syllable level

| Segmenter | P | R | F1 | Entities cut by a word boundary |
|---|---|---|---|---|
| VnCoreNLP | 80.15 | 88.90 | 84.30 | 0.26 % |
| **underthesea** | 80.05 | 88.68 | 84.15 | 0.84 % |
| none (raw syllables) | 76.00 | 86.91 | 81.09 | — |

Segmentation agreement between the two tools: word-level F1 97.9 %, 86.6 % of sentences identical; 98.0 % of underthesea's words are a single PhoBERT vocab entry (VnCoreNLP: 99.0 %), `<unk>` rate ≈ 0.02 %.

### Takeaways

- **The encoder is not harmed.** underthesea and VnCoreNLP are within noise on both tasks, frozen or fine-tuned (Δ ≤ 0.15 points).
- **Segmenting is still worth it**: skipping it costs −0.4 acc after fine-tuning and −3.2 NER F1.
- For **sequence labeling** the only measurable cost is word-boundary errors, not encoder quality: underthesea merges a function word or common noun into a following proper noun / number slightly more often (`từ_Anh`, `tối_8/3`, `trú_thôn Hạ_Lôi`), which cuts through ~0.8–1 % of gold entities vs ~0.2 % for VnCoreNLP. Use the same segmenter for training and inference, and project your labels onto its word boundaries.
- Note that `word_tokenize` normalises the text (`Hoà` → `Hòa`, drops the space before punctuation), so its output does not preserve the input string byte-for-byte — keep that in mind when aligning labels.

```python
from underthesea import word_tokenize
from transformers import AutoTokenizer, AutoModel

tok = AutoTokenizer.from_pretrained("vinai/phobert-base-v2")
model = AutoModel.from_pretrained("vinai/phobert-base-v2")

text = "Chàng trai 9X Quảng Trị khởi nghiệp từ nấm sò"
seg = word_tokenize(text, format="text")  # 'Chàng trai 9X Quảng_Trị khởi_nghiệp từ nấm sò'
out = model(**tok(seg, return_tensors="pt"))
```

Scripts and raw numbers: Kaggle kernel (GPU run, sentiment incl. fine-tuning) https://www.kaggle.com/code/rain1024/exp760-phobert-underthesea

**Comment by rain1024** [comment]
Closing — answered above with measurements. Feel free to reopen if you see a different result in your setup.
